### PR TITLE
[Core][Geometry] Adding pGetGeometryPart

### PR DESCRIPTION
**Pull request by tteschemacher**
Hi all,

I am adding here the interface to get the geometry parts with pointers.

Furthermore, the reason comes from the BrepClasses. Here, I added the corresponding function.

Thx for the review!

### CODE REVIEW

File: kratos/geometries/geometry.h [hunk 1]
```diff
@@ -1298,6 +1298,21 @@ class Geometry
...
+    /**
+    * @brief This function is necessary for composite geometries. It returns the
+    * pointer to a geometry part which is accessable with a certain index.
+    * @details This index
+    * is dependent on the derived implementation.
+    * @param Index of the geometry part. This index can be used differently
+    *        within the derived classes
+    * @return pointer to geometry, which is connected through the Index
+    */
+    virtual typename GeometryType::Pointer pGetGeometryPart(IndexType Index)
```
Comment (philbucher) [inline kratos/geometries/geometry.h]: you also need a `const` version

Comment (tteschemacher) [reply kratos/geometries/geometry.h]: I added it.

File: kratos/geometries/geometry.h [hunk 2]
```diff
@@ -676,6 +701,70 @@ class Geometry
...
+    ///@{
+
+    /**
+    * @brief Used for composite geometries. It returns the
+    *        the geometry part, corresponding to the Index.
+    * @param Index of the geometry part. This index can be used differently
+    *        within the derived classes.
+    * @return reference to corresponding geometry.
+     */
+    virtual GeometryType& GetGeometryPart(IndexType Index) const
```
Comment (tteschemacher) [inline kratos/geometries/geometry.h]: Should here also exist the const GetGeometryPart() const; plus GetGeometryPart()?
@philbucher @pooyan-dadvand

File: kratos/geometries/coupling_geometry.h [hunk 3]
```diff
@@ -192,20 +192,35 @@ template<class TPointType> class CouplingGeometry
...
+    * @brief This function returns the pointer of the geometry part
+    *        which is corresponding to the index.
+    *        Checks if index is available only in debug.
+    * @param Index: 0 -> master, all bigger than 0 -> slaves.
+    * @return pointer of geometry, corresponding to the index.
+    */
+    GeometryPointer pGetGeometryPart(IndexType Index) override
+    {
+        KRATOS_DEBUG_ERROR_IF(mpGeometries.size() <= Index) << "Index "
+            << Index << " out of range. Composite contains only of: "
```
Comment (pooyan-dadvand) [inline kratos/geometries/coupling_geometry.h]: Please revise the english

File: kratos/geometries/coupling_geometry.h [hunk 4]
```diff
@@ -192,20 +192,35 @@ template<class TPointType> class CouplingGeometry
...
+    * @brief This function returns the const pointer of the geometry part
+    *        which is corresponding to the index.
+    *        Checks if index is available only in debug.
+    * @param Index: 0 -> master, all bigger than 0 -> slaves.
+    * @return const pointer of geometry, corresponding to the index.
+    */
+    const GeometryPointer pGetGeometryPart(IndexType Index) const override
     {
         KRATOS_DEBUG_ERROR_IF(mpGeometries.size() <= Index) << "Index "
             << Index << " out of range. Composite contains only of: "
```
Comment (pooyan-dadvand) [inline kratos/geometries/coupling_geometry.h]: also here

**Review by pooyan-dadvand — COMMENTED** [review]
Now is ok for me except a few typos. @philbucher you may approve it if is ok for you.

File: kratos/geometries/coupling_geometry.h (hunk 3 above)
Comment (tteschemacher) [reply kratos/geometries/coupling_geometry.h]: I tried my best ;) hope it is better now.

File: kratos/geometries/brep_surface.h [hunk 5]
```diff
@@ -196,13 +199,97 @@ class BrepSurface
...
+    * @brief This function returns the pointer of the geometry
+    *        which is corresponding to the trim index.
+    *        Surface of the geometry is accessable with SURFACE_INDEX.
+    * @param Index: trim_index or SURFACE_INDEX.
+    * @return pointer of geometry, corresponding to the index.
+    */
+    GeometryPointer pGetGeometryPart(IndexType Index) override
+    {
+        const auto& const_this = *this;
+        return std::const_pointer_cast<GeometryType>(
```
Comment (philbucher) [inline kratos/geometries/brep_surface.h:216]: hm I am not sure abt this ...

**Review by philbucher — APPROVED** [review]

File: kratos/geometries/brep_surface.h (hunk 5 above)
Comment (tteschemacher) [reply kratos/geometries/brep_surface.h:216]: It's working. And so far it reduces the code significantly. I'll keep an eye on it to see how it performs in the future. If it shows off as bad we'll just duplicate the things...

**Comment by tteschemacher** [comment]
Thanks @pooyan-dadvand and @philbucher for your detailed review. I think we were able to make write something out of it.

**Pull request merged**